File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

# 최대 랜선수
N_MAX = 1_000_000
# 최대 길이
K_MAX = 10_000

# 입력 받기
k, n = map(int, input().split())

nums = [0] * K_MAX

# 각 랜선의 길이 입력
for i in range(k):
    nums[i] = int(input())

# 랜선 길이의 총합 계산
sum_length = sum(nums[:k])

# 시작, 끝, 최대 길이 초기화
start = 1
end = sum_length // n + 1
max_len = float('-inf')

# 이진 탐색을 이용한 최적의 랜선 길이 찾기
while start < end:
    # 중간 길이
    middle_len = start + (end - start) // 2
    # 남은 길이
    leftover = sum_length - (middle_len * n)
    # 남은 값들의 합
    leftover_sum = 0

    # 각 랜선의 길이를 이용하여 남은 값들의 합 계산
    for num in nums[:k]:
        leftover_sum += num % middle_len
        # 초과하면
        if leftover_sum > leftover:
            end = middle_len
            # 바깥쪽 반복문으로 이동
            break

    else:  # 만약 break가 발생하지 않으면 (모든 랜선이 해당 조건에 부합)
        # max_len 값을 업데이트
        max_len = max(middle_len, max_len)
        # 다음 반복을 위해 start를 조정
        start = middle_len + 1

# 최적의 랜선 길이 출력
print(max_len)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.rs
use std::io::{BufReader,BufRead,BufWriter,Write,stdin,stdout};
use std::iter::Iterator;
//최대 랜선수
const N_MAX: usize = 1_000_000;
//최대길이
const K_MAX: usize = 10_000;
fn main(){
  let mut reader= BufReader::new(stdin().lock());
  let mut writer= BufWriter::new(stdout().lock());
  let mut input= String::new();
  reader.read_line(&mut input).unwrap();
 let mut inputs= input.trim().split_whitespace().map(|x|x.parse::<u64>().unwrap());
 //랜선의 갯수k

 //필요한 랜선의 개수n
 let (k, n) = (inputs.next().unwrap() as usize, inputs.next().unwrap());

let mut nums= [0u64;K_MAX];

//각 랜선의 길이
for num in nums.iter_mut().take(k){
  input.clear();
  reader.read_line(&mut input).unwrap();
  //랜선의길이
  *num =input.trim().parse().unwrap();
  
}
let sum: u64 = nums.iter().take(k).sum();

//시작1
let mut start= 1u64;
//끝
let mut end= sum/n+1;
//최개길이
let mut max_len=u64::MIN;
//레이블 사용
//이진탐색
'outer: while start < end {
  //중간길이
  let middle_len = start + (end - start) / 2;
  //남은길이
  let leftover = sum - (middle_len * n);
  //남은 값들의합
  let mut leftover_sum = 0u64;
  for num in nums.iter().take(k) {
    //각 랜선의 길이를 이용하여 남은 값들의 합을 계산
    leftover_sum += *num % middle_len;
    //초과하면
    if leftover_sum > leftover {
        end = middle_len;
        //즉시 다음 레이블로 이동
        continue 'outer;
    }
   

   } 
    //max_len 값을 업데이트
    max_len = u64::max(middle_len, max_len);
    //다음 반복을 위해 start를 조정
    start = middle_len + 1;
}

writeln!(writer,"{max_len}").unwrap();

}